lsp/src/Language/LSP/VFS.hs

```diff
@@ -346,6 +353,45 @@ changeChars logger str start finish new = do
 
 -- ---------------------------------------------------------------------
 
+-- | A position, like a 'J.Position', but where the offsets in the line are measured in Unicode code points
+-- instead of UTF-16 code units.
+data CodePointPosition =
+  CodePointPosition
+    { -- | Line position in a document (zero-based).
+      _line      :: J.UInt
+      -- | Character offset on a line in a document in *code points* (zero-based).
+    , _character :: J.UInt
+    } deriving (Show, Read, Eq, Ord)
+
+-- | Given a virtual file, translate a 'CodePointPosition' in that file into a 'J.Position' in that file.
+--
+-- We need the file itself because this requires translating between code points and code units.
+codePointPositionToPosition :: VirtualFile -> CodePointPosition -> J.Position
+codePointPositionToPosition vFile (CodePointPosition cpl cpc) =
+  let text = _file_text vFile
+      lines = Rope.toTextLines text
+      -- Split at the given position in *code points*
+      (prefix, _) = ULines.splitAtPosition (ULines.Position (fromIntegral cpl) (fromIntegral cpc)) lines
+      -- Get the length of the prefix in *code units*
+      (Lines.Position cul cuc) = Lines.lengthAsPosition prefix
+  in J.Position (fromIntegral cul) (fromIntegral cuc)
```

Comment:
There are two tricks which can make it faster:
1. It's faster to split `Rope`, and convert `toTextLines` later. 
2. Line position is always preserved.

```suggestion
  let rope = _file_text vFile
      (_, rope') = Rope.splitAtLine (fromIntegral cpl)
      lines = Rope.toTextLines rope'
      -- Split at the given position in *code points*
      (prefix, _) = ULines.splitAt (fromIntegral cpc) lines
      -- Get the length of the prefix in *code units*
      cuc = Lines.length prefix
  in J.Position cpl (fromIntegral cuc)
```

This is still not quite satisfactory, because `Rope.toTextLines` materialises too much, while we really care about the first line of `rope'` only. Do we care about illegal positions, where column counter is bigger than a corresponsing line?

Reply:
I just pushed a version that just pulls out a single line and operates on that. I think that avoids the issue of materializing to much stuff. I'd appreciate a look.

> Do we care about illegal positions, where column counter is bigger than a corresponsing line?

We probably shouldn't crash, but it's not clear what we should do. Return `Nothing`, perhaps.

Reply:
My current implementation will return the position of the end of the line. I'll put a note to fix that.